src/common/hashtable.c

```diff
@@ -0,0 +1,351 @@
+}
+
+/*****************************************
+Name:  onf_ht_get_hash
+brief: get hash table
+return: 
+*****************************************/
+int onf_ht_get_hash(
+    const char* s, int key_len, 
+    const int num_buckets, const int attempt) 
+{
+    const int hash_a = onf_ht_hash(
+                        s, key_len, ONF_HT_PRIME_1, num_buckets);
+    const int hash_b = onf_ht_hash(
+                        s, key_len, ONF_HT_PRIME_2, num_buckets);
+    return (hash_a + (attempt * (hash_b + 1))) % num_buckets;
+}
+
+/*****************************************
+Name:  onf_ht_insert
+brief: insert into hash table
+return: 
+*****************************************/
+int onf_ht_insert(
+      onf_ht_hash_table* ht, int key_len,  
+      int val_len, const char* key, const char* value) 
+{
+    const int load = (ht->count * 100) / ht->size;
+    if (load > 70) {
+        if(SUCCESS != onf_ht_resize_up(ht))
+	{
+	    return E_FAIL;
+	}
+    }
+    
+    onf_ht_item* item = onf_ht_new_item(
+                         key_len, val_len,key, value);
+    if(NULL == item)
+    {
+       return E_ALLOC_FAILED;
+    }
+    int index = onf_ht_get_hash(item->key, 
+                   item->key_len,ht->size, 0);
+    onf_ht_item* cur_item = ht->items[index];
+    int i = 1;
+    
+    while (cur_item != NULL) {
+        if(cur_item != &ONF_HT_DELETED_ITEM) {
+            if (onf_key_compare(
+	          cur_item->key_len, key_len,
+		      cur_item->key, key)) {
+                onf_ht_del_item(cur_item);
+                ht->items[index] = item;
+                return SUCCESS;
+            }
+            else
+            {
+		index = onf_ht_get_hash(item->key, 
+		           item->key_len, ht->size, i);
+		cur_item = ht->items[index];
+		i++;
+            }
+        }
+    } 
+    ht->items[index] = item;
+    ht->count++;
+    return SUCCESS;
+}
+
+/*****************************************
+Name:  onf_htsearch
+brief: search key in hash table
+return: 
+*****************************************/
+char* onf_ht_search(onf_ht_hash_table* ht, const char* key, int key_len) {
+    int index = onf_ht_get_hash(key, key_len, ht->size, 0);
+    onf_ht_item* item = ht->items[index];
+    int i = 1;
+    while (item != NULL) {
+        if(item != &ONF_HT_DELETED_ITEM) {
+            if (onf_key_compare(
+	          item->key_len, key_len, item->key, key)) 
+            {
+                return item->value;
+            }
+        }
+        index = onf_ht_get_hash(key, key_len, ht->size, i);
+        item = ht->items[index];
+        i++;
+    } 
+    return NULL;
+}
+
+/*****************************************
+Name:  onf_ht_delete
+brief: delete from hash table
+return: 
+*****************************************/
+int onf_ht_delete(
+        onf_ht_hash_table* ht, 
+	const char* key, int key_len) 
+{
+    const int load = (ht->count * 100) / ht->size;
+    if (load > 10) {
+        if(SUCCESS != onf_ht_resize_down(ht))
+	{
+	    return E_FAIL;
```

Comment:
We should still continue to delete the element from bucket

Reply:
the allocation of memory has failed during resize. So, the application is not in a stable state. So better to fail.